### PR TITLE
refactor: mls feature config handling

### PR DESCRIPTION
**Pull request by typfel**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

Refactor how we handle the MLS feature configs: "mls" and "mlsMigration".

Previously these feature configs were either fetched by the `MLSMigrationRepository` or fetched when needed without persisting them (`UpdateSupportedProtocolsUseCase`) .

This became inflexible when I wanted to start updating the supported protocols in response to a  "mls" feature config update event. So in this PR I'm aligning the handling MLS feature configs with how other feature configs are handled, that is fetching is done by the `SyncFeatureConfigsUseCase` and all handling / persisting is done in each respective config handler. 

With these changed I was able to to: 
* Remove on demand fetching of the feature configs since I can rely on that the latest configuration has already been persisted.
* Delete `MLSMigrationRepository`, it's functionality is now handled by the config handlers.
* Update self supported protocols in response to mls feature config update events.

#### Test Coverage

- [x] I have added automated test to this contribution

### Notes

With these changed I think we can potentially also remove the "update supported protocols" slow sync step since it would any necessary update would be triggered by the MLS config handlers.

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/UserSessionScope.kt [hunk 1]
```diff
@@ -1030,9 +1047,11 @@ class UserSessionScope internal constructor(
 
     internal val mlsMigrationWorker get() =
         MLSMigrationWorkerImpl(
-            mlsMigrationRepository,
+            userConfigRepository,
+            featureConfigRepository,
+            MLSConfigHandler(userConfigRepository, updateSupportedProtocolsAndResolveOneOnOnes, userId),
```
Comment (mchenani) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/UserSessionScope.kt:1052]: suggestion: isn't better to define these as variables then we can re-use them?

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/UserSessionScope.kt [hunk 2]
```diff
@@ -1299,7 +1311,8 @@ class UserSessionScope internal constructor(
         get() = FeatureConfigEventReceiverImpl(
             GuestRoomConfigHandler(userConfigRepository, kaliumConfigs),
             FileSharingConfigHandler(userConfigRepository),
-            MLSConfigHandler(userConfigRepository, userId),
+            MLSConfigHandler(userConfigRepository, updateSupportedProtocolsAndResolveOneOnOnes, userId),
```
Comment (mchenani) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/UserSessionScope.kt:1314]: here we can re-use those variable for example.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/UserSessionScope.kt [hunk 3]
```diff
@@ -1562,7 +1573,8 @@ class UserSessionScope internal constructor(
             featureConfigRepository,
             GuestRoomConfigHandler(userConfigRepository, kaliumConfigs),
             FileSharingConfigHandler(userConfigRepository),
-            MLSConfigHandler(userConfigRepository, userId),
+            MLSConfigHandler(userConfigRepository, updateSupportedProtocolsAndResolveOneOnOnes, userId),
```
Comment (mchenani) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/UserSessionScope.kt:1576]: and here.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/featureConfig/SyncFeatureConfigsUseCase.kt [hunk 4]
```diff
@@ -63,7 +65,8 @@ internal class SyncFeatureConfigsUseCaseImpl(
             // TODO handle other feature flags and after it bump version in [SlowSyncManager.CURRENT_VERSION]
             guestRoomConfigHandler.handle(it.guestRoomLinkModel)
             fileSharingConfigHandler.handle(it.fileSharingModel)
-            mlsConfigHandler.handle(it.mlsModel)
+            mlsConfigHandler.handle(it.mlsModel, true)
```
Comment (mchenani) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/featureConfig/SyncFeatureConfigsUseCase.kt:68]: suggestion: could you please specify the variable name for 'true' value here for better readability?

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/mlsmigration/MLSMigrationManager.kt [hunk 5]
```diff
@@ -90,12 +89,16 @@ internal class MLSMigrationManagerImpl(
...
             kaliumLogger.d("Migration needs to be updated: $lastMlsMigrationCheckHasPassed")
             if (lastMlsMigrationCheckHasPassed) {
-                kaliumLogger.d("Updating migration config")
-                mlsMigrationRepository.value.fetchMigrationConfiguration()
-                    .onSuccess { timestampKeyRepository.value.reset(TimestampKeys.LAST_MLS_MIGRATION_CHECK) }
-                    .flatMap {
-                        mlsMigrationWorker.value.runMigration()
-                        Either.Right(Unit)
+                kaliumLogger.d("Running mls migration")
+                println("running migration")
```
Comment (mchenani) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/mlsmigration/MLSMigrationManager.kt:93]: question: missed debug logs? or if it's needed convert to logger?

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/mlsmigration/MLSMigrationManager.kt [hunk 6]
```diff
@@ -90,12 +89,16 @@ internal class MLSMigrationManagerImpl(
...
+                mlsMigrationWorker.value.runMigration()
+                    .onSuccess {
+                        println("done migration")
```
Comment (mchenani) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/mlsmigration/MLSMigrationManager.kt:96]: here too.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/mlsmigration/MLSMigrationWorker.kt [hunk 7]
```diff
@@ -17,46 +17,52 @@
...
-                kaliumLogger.i("MLS migration is not enabled")
-                Either.Right(Unit)
-            }
-        } ?: Either.Right(Unit)
+            } ?: Either.Right(Unit)
+        }
+
+    private suspend fun syncMigrationConfigurations(): Either<CoreFailure, Unit> =
+        featureConfigRepository.getFeatureConfigs().flatMap { configurations ->
+            mlsConfigHandler.handle(configurations.mlsModel, true)
```
Comment (mchenani) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/mlsmigration/MLSMigrationWorker.kt:63]: suggestion: add variable name for 'true' value.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/sync/receiver/FeatureConfigEventReceiver.kt [hunk 8]
```diff
@@ -84,7 +86,8 @@ internal class FeatureConfigEventReceiverImpl internal constructor(
     private suspend fun handleFeatureConfigEvent(event: Event.FeatureConfig): Either<CoreFailure, Unit> =
         when (event) {
             is Event.FeatureConfig.FileSharingUpdated -> fileSharingConfigHandler.handle(event.model)
-            is Event.FeatureConfig.MLSUpdated -> mlsConfigHandler.handle(event.model)
+            is Event.FeatureConfig.MLSUpdated -> mlsConfigHandler.handle(event.model, false)
```
Comment (mchenani) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/sync/receiver/FeatureConfigEventReceiver.kt:89]: suggestion: add arg name for 'false' value.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/sync/receiver/FeatureConfigEventReceiver.kt [hunk 9]
```diff
@@ -84,7 +86,8 @@ internal class FeatureConfigEventReceiverImpl internal constructor(
...
+            is Event.FeatureConfig.MLSMigrationUpdated -> mlsMigrationConfigHandler.handle(event.model, false)
```
Comment (mchenani) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/sync/receiver/FeatureConfigEventReceiver.kt:90]: suggestion: add arg name for 'false' value.

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/mlsmigration/MLSMigrationManagerTest.kt [hunk 10]
```diff
@@ -51,13 +51,16 @@ class MLSMigrationManagerTest {
...
-                .withLastMLSMigrationChecResetSucceeds()
+                .withRunMigrationSucceeds()
+                .withLastMLSMigrationCheckResetSucceeds()
                 .arrange()
 
             arrangement.incrementalSyncRepository.updateIncrementalSyncState(IncrementalSyncStatus.Live)
+            advanceUntilIdle()
             yield()
 
+            println("checking migration")
```
Comment (mchenani) [inline logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/mlsmigration/MLSMigrationManagerTest.kt:62]: missed log to remove?

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/featureConfig/FeatureConfigMapper.kt [hunk 11]
```diff
@@ -152,3 +153,17 @@ class FeatureConfigMapperImpl : FeatureConfigMapper {
...
+fun MLSMigrationModel.toEntity(): MLSMigrationEntity =
+    MLSMigrationEntity(
+        status = status.equals(Status.ENABLED),
+        startTime = startTime,
+        endTime = endTime
+    )
+
+fun MLSMigrationEntity.toModel(): MLSMigrationModel =
+    MLSMigrationModel(
+        status = if (status) Status.ENABLED else Status.DISABLED,
```
Comment (mchenani) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/featureConfig/FeatureConfigMapper.kt:166]: question: isn't the status itself boolean?

**Review by mchenani — APPROVED** [review]
🤘 LGTM, left some comments and questions.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/UserSessionScope.kt (hunk 1 above)
Comment (typfel) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/UserSessionScope.kt:1052]: Define the config handlers as separate variables and re-use them you mean?

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/mlsmigration/MLSMigrationManager.kt (hunk 5 above)
Comment (typfel) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/mlsmigration/MLSMigrationManager.kt:93]: I'll delete them

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/mlsmigration/MLSMigrationManager.kt (hunk 6 above)
Comment (typfel) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/mlsmigration/MLSMigrationManager.kt:96]: 💥

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/featureConfig/FeatureConfigMapper.kt (hunk 11 above)
Comment (typfel) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/data/featureConfig/FeatureConfigMapper.kt:166]: No, it's an enum for some reason.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/UserSessionScope.kt (hunk 1 above)
Comment (typfel) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/UserSessionScope.kt]: Done

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/UserSessionScope.kt (hunk 2 above)
Comment (typfel) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/UserSessionScope.kt]: Done

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/UserSessionScope.kt (hunk 3 above)
Comment (typfel) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/UserSessionScope.kt]: Done

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/featureConfig/SyncFeatureConfigsUseCase.kt (hunk 4 above)
Comment (typfel) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/featureConfig/SyncFeatureConfigsUseCase.kt]: Done

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/mlsmigration/MLSMigrationWorker.kt (hunk 7 above)
Comment (typfel) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/mlsmigration/MLSMigrationWorker.kt]: Done, and the value was actually wrong.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/sync/receiver/FeatureConfigEventReceiver.kt (hunk 8 above)
Comment (typfel) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/sync/receiver/FeatureConfigEventReceiver.kt]: Done

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/sync/receiver/FeatureConfigEventReceiver.kt (hunk 9 above)
Comment (typfel) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/sync/receiver/FeatureConfigEventReceiver.kt]: Done

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/mlsmigration/MLSMigrationManagerTest.kt (hunk 10 above)
Comment (typfel) [reply logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/mlsmigration/MLSMigrationManagerTest.kt]: Deleted

**Review by vitorhugods — APPROVED** [review]

**Pull request merged**